subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java

```diff
@@ -60,51 +102,83 @@ public void run() {
 
             private void doRun() {
                 while (true) {
+                    DatagramPacket packet;
                     long lockId;
                     try {
-                        lockId = communicator.receive();
+                        packet = communicator.receive();
+                        lockId = communicator.decodeLockId(packet);
                     } catch (GracefullyStoppedException e) {
                         return;
                     }
+
                     lock.lock();
-                    Runnable action;
-                    try {
-                        action = contendedActions.get(lockId);
-                        if (action == null) {
-                            //received access request for lock that is already closed
-                            continue;
+                    ContendedAction contendedAction = contendedActions.get(lockId);
+                    if (contendedAction == null) {
+                        acceptConfirmationAsLockRequester(lockId, packet.getPort());
+                    } else {
+                        if (!contendedAction.running) {
+                            startLockReleaseAsLockHolder(contendedAction);
                         }
-                    } finally {
-                        lock.unlock();
+                        communicator.confirmUnlockRequest(packet);
                     }
-                    action.run();
+                    lock.unlock();
                 }
             }
         };
     }
 
+    private void startLockReleaseAsLockHolder(ContendedAction contendedAction) {
+        unlockActionExecutor.execute(contendedAction.action);
+        contendedAction.running = true;
```

Comment:
I understand `unlockActionExecutor` will do it asynchronously, but maybe putting `running=true` before would be less puzzling.